# PostgreSQL & RDKit Cartridge & Docker

* RDKitì€ PostgreSQLìš© Database Cartridgeë¥¼ ì œê³µí•œë‹¤.

## Docker ê¸°ë°˜ PostgreSQL ì»¨í…Œì´ë„ˆë¥¼ ë„ìš°ê³  í•´ë‹¹ í™˜ê²½ì—ì„œ ChemBL ë°ì´í„°ë² ì´ìŠ¤ë¥¼ ìƒì„±í•˜ëŠ” ë°©ë²•

1. Docker image ë‹¤ìš´ ë° ì»¨í…Œì´ë„ˆ ì‹¤í–‰
Git Bashì—ì„œ ì„¤ì •í•¨

```
$ docker run --name dockerpostgres -p 5432:5432 -e POSTGRES_PASSWORD=(postgreSQLë¹„ë°€ë²ˆí˜¸) -d mcs07/postgres-rdkit

or

$ docker run --name some-postgres -e POSTGRES_PASSWORD=mysecretpassword -d postgres
ì•„ë˜ ì½”ë“œëŠ” rdkit ì—†ëŠ”ê±´ê°€
```
    > --name: ì»¨í…Œì´ë„ˆ ì´ë¦„
    > dockerpostgres: NAMES(ì •í•  ìˆ˜ ìˆìŒ)
    > -d: ë°±ê·¸ë¼ìš´ë“œë¡œ ì»¨í…Œì´ë„ˆ ì‹¤í–‰
    > mcs07/postgres-rdkit: ë¶ˆëŸ¬ì˜¬ docker image ì´ë¦„


```
$ docker ps
```
> Dockerì˜ ì •ë³´ë¥¼ ë³¼ìˆ˜ ìˆìŒ(CONTAINER ID, IMAGE, COMMAND, NAMES ...)

2. ì»¨í…Œì´ë„ˆ ì ‘ì† ë° ChEMBL ë°ì´í„° ë‹¤ìš´ë¡œë“œ

```
$ docker exec -it dockerpostgres /bin/bash
```
> ë„ì»¤ì˜ PostgreSQL ì»¨í…Œì´ë„ˆë¡œ ì ‘ì†í•˜ê²Œ ëœë‹¤.

```
$ wget ftp://ftp.ebi.ac.uk/pub/databases/chembl/ChEMBLdb/latest/chembl_30_postgresql.tar.gz
```
> ìœ„ ì½”ë“œë¥¼ ì…ë ¥í•˜ë©´ ë‹¤ìš´ì´ ë˜ëŠ”ë° ì•ˆëœë‹¤ë©´ ì•„ë˜ ì½”ë“œë¡œ wgetì„ ì„¤ì¹˜í•´ì•¼í•¨(ë„ì»¤ ì»¨í…Œì´ë„ˆì—ì„œ ì‹¤í–‰)

```
apt-get update && apt-get install wget
```
> (ë„ì»¤ ì»¨í…Œì´ë„ˆì—ì„œ ì‹¤í–‰)

3. PostgreSQL Shell ì ‘ì† ë° ë°ì´í„° í…Œì´ë¸” ìƒì„±

```
$ psql -U postgres
```
> (ë„ì»¤ ì»¨í…Œì´ë„ˆì—ì„œ ì‹¤í–‰í•˜ë©´ PostgreSQL Shellë¡œ ì ‘ì†í•˜ê²Œ ë¨)

```
create database chembl_30;
```
4. ë„ì»¤ ì»¨í…Œì´ë„ˆë¡œ ë¹ ì ¸ë‚˜ì™€ì„œ ChEMBL_30 íŒŒì¼ ì••ì¶•í•´ì œ ë° dmp íŒŒì¼ ë°ì´í„° ë² ì´ìŠ¤ì— ì €ì¥ (ì‹œê°„ ì†Œìš”)

```bash
$ tar -vxf  chembl_30_postgresql.tar.gz # ì••ì¶œ í’€ê¸°
$ cd chembl_30/chemb_30_postgresql
$ pg_restore --no-owner -h localhost -p 5432 -U postgres -d chembl_30 chembl_30_postgresql.dmp
```

5. ìƒì„±ëœ Chembl_30 DBì— ì ‘ê·¼ ë° í…Œì´ë¸” ìƒì„±

```
psql -U postgres chembl_30
```
* Connect to the database, install the cartridge, and create the schema that weâ€™ll use:

```sql
$ create extension if not exists rdkit;
$ create schema rdk;
```
* Create the molecules and build the substructure search index:

```sql
$ select * into rdk.mols from (select molregno,mol_from_ctab(molfile::cstring) m  from compound_structures) tmp where m is not null;
$ create index molidx on rdk.mols using gist(m);
$ alter table rdk.mols add primary key (molregno);
```

* Create some fingerprints and build the similarity search index:

```sql
$ select molregno,torsionbv_fp(m) as torsionbv,morganbv_fp(m) as mfp2,featmorganbv_fp(m) as ffp2 into rdk.fps from rdk.mols;
$ create index fps_ttbv_idx on rdk.fps using gist(torsionbv);
$ create index fps_mfp2_idx on rdk.fps using gist(mfp2);
$ create index fps_ffp2_idx on rdk.fps using gist(ffp2);
$ alter table rdk.fps add primary key (molregno);
```
* Here is a group of the commands used here (and below) in one block so that you can just paste it in at the psql prompt:

```sql
create or replace function get_mfp2_neighbors(smiles text)
returns table(molregno bigint, m mol, similarity double precision) as
$$
select molregno,m,tanimoto_sml(morganbv_fp(mol_from_smiles($1::cstring)),mfp2) as similarity
from rdk.fps join rdk.mols using (molregno)
where morganbv_fp(mol_from_smiles($1::cstring))%mfp2
order by morganbv_fp(mol_from_smiles($1::cstring))<%>mfp2;
$$ language sql stable ;
```

# PostgreSQL & Python
* pythonë²„ì „ì— ë§ëŠ” pipì„ ë‹¤ì‹œ ì„¤ì¹˜

```bash
$ apt-get remove python3-pip
$ apt-get install python3-pip
```

## 1. PostgreSQL Docker ì»¨í…Œì´ë„ˆì— í•„ìš” í”„ë¡œê·¸ë¨ ì„¤ì¹˜

```bash
$ apt-get update
$ apt-get install libpq-dev
```
## 2. psycopg2 ì„¤ì¹˜

```
$ pip3 install psycopg2
$ pip3 install psycopg2-binary
```
> conda ê°€ìƒí™˜ê²½ì—ì„œ ì‚¬ìš©

bash: pip: command not found ë¼ëŠ” ë¬¸êµ¬ê°€ ë‚˜ì˜¤ë©´ ì•„ë˜ ì½”ë“œ ì…ë ¥

```
$ apt-get install python3-pip
```
> pipê°€ ì•ˆ ê¹”ë ¤ìˆì„ë•Œ ë‹¤ìš´ë¡œë“œí•  ìˆ˜ ìˆìŒ.

> pipëŠ” íŒŒì´ì¬ìœ¼ë¡œ ì‘ì„±ëœ íŒ¨í‚¤ì§€ ì†Œí”„íŠ¸ì›¨ì–´ë¥¼ ì„¤ì¹˜ Â· ê´€ë¦¬í•˜ëŠ” íŒ¨í‚¤ì§€ ê´€ë¦¬ ì‹œìŠ¤í…œì´ë‹¤.

> python3ë²„ì „ì˜ ê²½ìš° íŒŒì´ì¬ ëª…ë ¹ì–´ë„ python3ë¥¼ ì‚¬ìš©í•˜ëŠ”ë° pipë„ ë§ˆì°¬ê°€ì§€ë¡œ pip3 ëª…ë ¹ì–´ë¥¼ ì‚¬ìš©í•œë‹¤.

## 3. pip list ì €ì¥
pip freeze > requirements.txt



## 4. ì°¸ê³ í•œ ì‚¬ì´íŠ¸
["AIë¡œ ì‹ ì•½ ë¬¼ì§ˆê³¼ íƒ€ê¹ƒ ë°œêµ´í•˜ëŠ” íŒ¨í‚¤ì§€ë¡œ ìŠ¹ë¶€"_ê¸°ì‚¬](http://www.hitnews.co.kr/news/articleView.html?idxno=35920)

[ì‹ ì•½ê°œë°œì—ì„œ ë”¥ ëŸ¬ë‹ì˜ ë¶€ìƒ](https://www.ibric.org/myboard/read.php?Board=report&id=3349)

[ë¨¸ì‹ ëŸ¬ë‹ ê¸°ë°˜ ë¶„ì í™œì„± ì˜ˆì¸¡ ì •í™•ë„ê°€ ë‚®ì€ ì´ìœ ](https://novelism.co.kr/197?category=930849)

[fragment ê¸°ë°˜ ì•½ë¬¼ ê°€ìƒíƒìƒ‰](https://novelism.co.kr/188?category=930849)

[PostgreSQL & RDKit Cartridge & Docker](https://velog.io/@decom0405/PostgreSQL-RDKit-Cartridge-1)

[ë„ì»¤ ê°€ìƒí™” ë¦¬ëˆ…ìŠ¤ í™˜ê²½ì—ì„œ íŒŒì´ì¬ ë²„ì „ ë³€ê²½(Ubuntu 18.04)](https://hobbylists.tistory.com/entry/%EB%8F%84%EC%BB%A4-%EA%B0%80%EC%83%81%ED%99%94-%EB%A6%AC%EB%88%85%EC%8A%A4-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C-%ED%8C%8C%EC%9D%B4%EC%8D%AC-%EB%B2%84%EC%A0%84-%EB%B3%80%EA%B2%BDUbuntu-1804)

[ë¦¬ëˆ…ìŠ¤ ê¸°ë³¸ëª…ë ¹ì–´](https://www.mireene.com/webimg/linux_tip1.htm)

[Docker compose ì‚¬ìš©ë²•1](https://engineer-mole.tistory.com/221)  

[Docker compose ì‚¬ìš©ë²•2](https://darrengwon.tistory.com/793)

[Docker compose ë¥¼ ì´ìš©í•´ì„œ ë³µì¡í•œ ë„ì»¤ ì»¨í…Œì´ë„ˆë¥¼ ì œì–´í•˜ê¸°_ìœ íŠœë¸Œ](https://www.youtube.com/watch?v=EK6iYRCIjYs)

[ë„ì»¤ë¥¼ ì´ìš©í•œ íŒŒì´ì¬ ê°œë°œ](https://inaj012.medium.com/%EB%8F%84%EC%BB%A4%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%B4-python-%EA%B0%9C%EB%B0%9C-f6f1af58bf6)  

[VS codeì— Docker ì—°ë™í•˜ê¸°](https://89douner.tistory.com/123)

[VS codeë¥¼ í™œìš©í•˜ì—¬ ì›ê²© ì„œë²„ì˜ docker containerì— ì ‘ì†í•˜ëŠ” ë°©ë²•](https://seokhyun2.tistory.com/48)

[Docker image ìƒì„± ë° ë°°í¬í•˜ê¸°](https://hororolol.tistory.com/559)

[PostgreSQL & Python](https://velog.io/@decom0405/PostgreSQL-Python)

[SQL Examples](https://chembl.gitbook.io/chembl-interface-documentation/frequently-asked-questions/schema-questions-and-sql-examples#retrieve-all-the-bioactivity-data-for-bacterial-targets)  

[The RDKit database cartridge](https://www.rdkit.org/docs/Cartridge.html)

[íŒŒì´ì¬ 3.8ì—ì„œ RDKit ì„¤ì¹˜](https://mopipe.tistory.com/109)

[SPE2Vec ë¬¸ì„œ ì„¤ëª…](https://xinhaoli74.github.io/SmilesPE/spe2vec/)

[ìºê¸€_Drug Design with Small Molecule SMILES](https://www.kaggle.com/code/art3mis/220221-getting-started-with-smiles)

[ì½”ë“œìŠ¤í…Œì´ì¸  cp1 í”¼ë“œë°±](https://codestates.notion.site/AIBootcamp-9-10-Code-States-Project-1-d1a05d1bc6914f91a462963cfba33aae#583f0053f8624a8984e7aa7fa7e559b9)

* ì°¸ê³ ë…¼ë¬¸

[J. Chem. Inf. Model. 2021, 61, 4, 1560â€“1569](https://pubs.acs.org/doi/10.1021/acs.jcim.0c01127)


## 5. ì¶”ê°€ ê³µë¶€
ë…¼ë¬¸  
[DeepPLA: a novel deep learning-based model for protein-ligand binding affinity
prediction](https://www.biorxiv.org/content/10.1101/2021.12.01.470868v1.full.pdf)  

ì‹ ì•½ì„ ì‹œì¥ì— ì¶œì‹œí•˜ëŠ” ê²ƒì€ ìœ„í—˜í•˜ê³ , ì‹œê°„ì´ ë§ì´ ê±¸ë¦¬ë©°, ë¹„ìš©ì´ ë§ì´ ë“œëŠ” ê²ƒì´ íŠ¹ì§•ì´ë‹¤.
ì‹¤ì œë¡œ 2018ë…„ì— ë³´ê³ ëœ ë°”ì™€ ê°™ì´ ì‹ ì•½ ê°œë°œì—ëŠ” 10~20ë…„, 20ì–µ ë‹¬ëŸ¬ ì´ìƒì´ ì†Œìš”ë  ìˆ˜ ìˆë‹¤. ì´ ë‹¨ê³„ëŠ” ì¼ë°˜ì ìœ¼ë¡œ ìˆ˜ë°± ìˆ˜ì²œ ê°œì˜ í™”í•©ë¬¼ì„ í¬í•¨í•˜ì§€ë§Œ, ì•½ 8%ë§Œì´ ì²´ì™¸ ë° ì²´ë‚´ ì„ìƒ ì—°êµ¬ ë‹¨ê³„ì— ì§„ì…í•  ìˆ˜ ìˆë‹¤[4].
ëª¨ë“  ì•½ë¬¼ ìš©ë„ ë³€ê²½ ë°©ë²• ê°€ìš´ë°, ì˜ì•½í’ˆ í™”í•©ë¬¼ ë¼ì´ë¸ŒëŸ¬ë¦¬ë¥¼ ì„ ë³„í•˜ê³  ì•½ë¬¼ í‘œì  ìƒí˜¸ ì‘ìš©(DTI)ì„ ì‹ë³„í•˜ëŠ” ì‹¤ë¦¬ì½” ê³„ì‚° ê¸°ë°˜ ë°©ë²•ì€ ë†’ì€ ì»´í“¨íŒ… ì•„í‚¤í…ì²˜ì˜ ë°œì „ê³¼ ë¨¸ì‹  ëŸ¬ë‹ì˜ ë°œì „ ë•ë¶„ì— ì ì  ë” ë§ì€ ê´€ì‹¬ì„ ì–»ê³  ìƒë‹¹í•œ ë°œì „ì„ ì´ë£¸.

ì¼ë°˜ì ìœ¼ë¡œ ì˜ì•½í’ˆ ì—°êµ¬ ë° ê°œë°œì˜ ì „ì²´ í”„ë¡œì„¸ìŠ¤ëŠ” (i) ì˜ì•½í’ˆ ë°œê²¬, (ii) ì„ìƒ ì—°êµ¬, (iii) ì„ìƒ ì‹œí—˜, (iv) FDA ê²€í†  ë° (v) ì‹œíŒ í›„ ì•ˆì „ì„± ê°ì‹œë¥¼ í¬í•¨í•œ ì„œë²Œ í•µì‹¬ ë‹¨ê³„[3]ë¡œ ë³µì¡í•˜ë‹¤. íŠ¹íˆ ì•½ë¬¼ ë°œê²¬ì€ ë‹¨ë°±ì§ˆê³¼ ê°™ì€ ì¶©ì¡±ë˜ì§€ ì•Šì€ ì§ˆë³‘ì˜ ëŒ€ìƒì„ ì‹ë³„í•˜ëŠ” ê²ƒìœ¼ë¡œ ì‹œì‘í•˜ì—¬ ëŒ€ìƒê³¼ íš¨ìœ¨ì ì´ê³  ì•ˆì „í•˜ê²Œ ìƒí˜¸ ì‘ìš©í•  ìˆ˜ ìˆëŠ” ìœ ë§í•œ í™”í•©ë¬¼ì„ ìƒì„±í•˜ê³  ìµœì í™”í•˜ëŠ” ì²« ë²ˆì§¸ ë‹¨ê³„ì´ë‹¤.

ì•½ë¬¼ í‘œì  ìƒí˜¸ ì‘ìš©(DTI)
ëª©ì : í‘œì  ë‹¨ë°±ì§ˆê³¼ ì•½ë¬¼ í™”í•©ë¬¼ ì‚¬ì´ì˜ ë‹¨ë°±ì§ˆ ë¦¬ê°„ë“œ ìƒí˜¸ ì‘ìš©ì˜ ê°•ë„ë¥¼ ì¸¡ì •í•˜ëŠ” ê²°í•© ì¹œí™”ì„±ì„ ì—°êµ¬í•˜ëŠ” ê²ƒ
* ê²°í•© ì¹œí™”ì„±ì€ ì¼ë°˜ì ìœ¼ë¡œ inhibition(ğ¾ğ‘–), dissociation(ğ¾ğ‘‘) ë° half-maximal inhibitory(ğ¼ğ¶50)ë¥¼ ë‚˜íƒ€ë‚´ëŠ” ë‹¤ìŒê³¼ ê°™ì€ ìƒìˆ˜ì˜ í˜•íƒœë¡œ í‘œí˜„
* ë‹¨ë°±ì§ˆ-ë¦¬ê°„ë“œ ìƒí˜¸ì‘ìš©ì„ ì‹ë³„í•˜ë ¤ëŠ” ê¸°ì¡´ì˜ ë°©ë²•ì€ ë¶„ì/ìƒë¬¼í•™ì  êµ¬ì¡°ì˜ ë³µì¡í•œ ì–‘ì í™”í•™ ê³„ì‚°ì„ í¬í•¨
  * ê¸°ì¡´ ë°©ë²•ì€ ê¸ì •ì ì¸ ì•½ë¬¼-í‘œì  ìŒì„ ë°œê²¬í•˜ê¸° ìœ„í•´ ê¸°ì¡´ ë°ì´í„°ë² ì´ìŠ¤ì—ì„œ í•™ìŠµí•  ìˆ˜ ìˆëŠ” ë°©ë²•ì´ ë  ìˆ˜ ì—†ë‹¤
  * ì§€ë‚œ 10ë…„ ë™ì•ˆ ëœë¤ í¬ë ˆìŠ¤íŠ¸(RF) ê¸°ë°˜ ì•Œê³ ë¦¬ë“¬[8][9], SimBoost[10] ë° ChemBoost[11]ì™€ ê°™ì€ ìˆ˜ë°±ë§Œ ë¦¬ê°„ë“œì™€ ë‹¨ë°±ì§ˆì—ì„œ DTIë¥¼ ì‹ë³„í•˜ê¸° ìœ„í•´ ë‹¤ì–‘í•œ ê¸°ê³„ í•™ìŠµ ê¸°ë°˜ ëª¨ë¸ì´ ê°œë°œ
  * ë”¥ ëŸ¬ë‹ ê¸°ë°˜ ê¸°ë²•ì˜ ì¶œí˜„ê³¼ ê·¸ê²ƒì˜ ì„±ê³µì ì¸ ì‘ìš©ì€ ì»´í“¨í„° ë¹„ì „ì´ë‚˜ ì–¸ì–´ ì²˜ë¦¬ì™€ ê°™ì€ ì‘ìš© í”„ë¡œê·¸ë¨ì„ ë„˜ì–´ ì‹ ì•½ì„ ë°œê²¬í•˜ëŠ” ìœ ë§í•œ ë°©ë²•ì„ ê°œì²™í•¨
    * ì˜ˆ) 
    * MFDR[12]ì€ í™”í•™ êµ¬ì¡° ë° ë‹¨ë°±ì§ˆ ì‹œí€€ìŠ¤ì—ì„œ ìë™ ì¸ì½”ë”ë¡œ ê¸°ëŠ¥ì„ ì–»ì€ ë‹¤ìŒ SVM ëª¨ë¸ì„ ì‚¬ìš©í•˜ì—¬ DTIë¥¼ 0-1 ë¶„ë¥˜ ë¬¸ì œë¡œ ì˜ˆì¸¡
    * DeepDTA[13]ëŠ” í™”í•©ë¬¼ê³¼ ë‹¨ë°±ì§ˆì˜ ì‹œí€€ìŠ¤ë¥¼ ì •ë³´ ëª¨ë“ˆë¡œ í‘œí˜„í•˜ê¸° ìœ„í•´ ë‘ ê°œì˜ ë‹¤ë¥¸ CNN ëª¨ë“ˆì„ ë³„ë„ë¡œ ì ìš©
    * ì •ë³´ ëª¨ë“ˆì˜ ê²°ê³¼ì  íŠ¹ì§•ì€ ë‹¨ë°±ì§ˆ-ë¦¬ê°„ë“œ ê²°í•© ì¹œí™”ì„±ì„ ì˜ˆì¸¡í•˜ê¸° ìœ„í•´ ì™„ì „íˆ ì—°ê²°ëœ 3ê°œì˜ ë ˆì´ì–´ë¥¼ ì…ë ¥
      * Deepì—ì„œ ê°„ë‹¨í•œ ë¼ë²¨ ì¸ì½”ë”© ë°©ë²•ì˜ ì‚¬ìš©ì›ì‹œ ì…ë ¥ ì‹œí€€ìŠ¤(ì¦‰, í•´ë‹¹í•˜ëŠ” ì¸ì½”ë”©ëœ ì •ìˆ˜ë¥¼ ì‚¬ìš©í•˜ì—¬ ì›ì‹œ ì‹œí€€ìŠ¤ì—ì„œ ê¸°í˜¸ë¥¼ ë‚˜íƒ€ëƒ„)ë¥¼ í¬í•¨í•˜ê¸° ìœ„í•œ DTAëŠ” ì›ì‹œ ì‹œí€€ìŠ¤ì— ëŒ€í•œ ë§ì€ ì •ë³´ë¥¼ ìƒì„ ìˆ˜ ìˆë‹¤.
      * OnionNet[15]ì€ ë‹¨ë°±ì§ˆê³¼ ì•½ë¬¼ ë¶„ìì˜ 3D êµ¬ì¡°ë¥¼ í™œìš©í•˜ì—¬ ì•½ë¬¼ í‘œì  ê²°í•© ì¹œí™”ì„±[16]ì„ ì˜ˆì¸¡í•˜ì—¬ ì •ë³´ ì†ì‹¤ ë¬¸ì œë¥¼ í¬ê²Œ ì¤„ì¸ë‹¤. 
      * ë¶ˆì¶©ë¶„í•œ 3D ë‹¨ë°±ì§ˆ êµ¬ì¡° ë°ì´í„°[17]ë¡œ ì¸í•´ ì´ëŸ¬í•œ ëª¨ë¸ì˜ ì‹¤ìš©ì„±ê³¼ ì •í™•ì„±ì€ ì œí•œëœë‹¤. ë‹¨ë°±ì§ˆì— ëŒ€í•œ ì •í™•í•œ 3D êµ¬ì¡° ë°ì´í„°ëŠ” ê°€í˜¹í•˜ê³  ê·¹ë‹¨ì ì¸ ì¡°ê±´ì—ì„œ ê³ ê¸‰ ì‹¤í—˜ ë°©ë²•ì´ í•„ìš”í•˜ê¸° ë•Œë¬¸ì— ì–»ê¸° ì–´ë µë‹¤[18].

* ë³¸ ë…¼ë¬¸ì—ì„œëŠ” ë‹¨ë°±ì§ˆ 1D ì‹œí€€ìŠ¤ì™€ 1D ë¦¬ê°„ë“œ SMILES ë¬¸ìì—´ì„ ì‚¬ìš©í•˜ëŠ” ë‹¨ë°±ì§ˆ-ë¦¬ê°„ë“œ ê²°í•© ì¹œí™”ë„ ì˜ˆì¸¡ì„ ìœ„í•œ í˜ì‹ ì ì¸ ë”¥ ëŸ¬ë‹ ê¸°ë°˜ ëª¨ë¸ì¸ DeepPLAë¥¼ ì œì•ˆ
  * Simplified Molecular Input Line Entry Systemì˜ ì•½ì SMILESëŠ” í™”í•™ ê·œì¹™ì— ê¸°ì´ˆí•œ ë¶„ì êµ¬ì¡°ë¥¼ í‘œí˜„í•˜ê¸° ìœ„í•œ ì˜ ì•Œë ¤ì§„ ì½¤íŒ©íŠ¸ ì„ í˜• í‘œê¸°ë²•ì´ë‹¤.
    * Mol2Vec[20] ë° ProSE[21]ì˜ ì‚¬ì „ í›ˆë ¨ëœ ëª¨ë¸ì„ ê°ê° ì‚¬ìš©í•˜ì—¬ ë¦¬ê°„ë“œ SMILES ë¬¸ìì—´ê³¼ ë‹¨ë°±ì§ˆ ì‹œí€€ìŠ¤ë¥¼ ìˆ«ì ë²¡í„°ë¡œ ë‚´ì¥í•œ ë‹¤ìŒ ë‹¨ë°±ì§ˆê³¼ ë¦¬ê°„ë“œì˜ ë‚´ì¥ ìˆ«ì ë²¡í„°ë¥¼ Head CNN ëª¨ë“ˆê³¼ ResNet ê¸°ë°˜ CNN ëª¨ë“ˆì— ê³µê¸‰í•˜ì—¬ ê°ê° ë‹¨ë°±ì§ˆê³¼ ë¦¬ê°„ë“œ ì‹œí€€ìŠ¤ë¥¼ ì¸ì½”ë”©í•œë‹¤.
    * ì¸ì½”ë”©ëœ í‘œí˜„ì€ ë²¡í„°ë¡œ ì—°ê²°ë˜ê³  ì–‘ë°©í–¥ LSTM(ë¡± ë‹¨ê¸° ë©”ëª¨ë¦¬) ê³„ì¸µì— ì¶”ê°€ë¡œ ê³µê¸‰ë˜ë©°, ì´ì–´ì„œ ì„¸ ê°œì˜ ì™„ì „íˆ ì—°ê²°ëœ ê³„ì¸µì´ ì´ì–´ì§„ë‹¤. DeepPLAì˜ ì¶œë ¥ì€ ì—°ì† ê°’ì´ë©°, ë¦¬ê°„ë“œì™€ ë‹¨ë°±ì§ˆì˜ ì…ë ¥ ìŒì˜ ê²°í•© ì¹œí™”ì„±ì„ ë‚˜íƒ€ë‚¸ë‹¤. ìš°ë¦¬ëŠ” BindingDB ë°ì´í„° ì„¸íŠ¸[22]ë¥¼ ì‚¬ìš©í•˜ì—¬ DeepPLA ëª¨ë¸ì„ í›ˆë ¨í•˜ê³  ë°”ì¸ë”© ì¹œí™”ë„ì— ëŒ€í•œ ì„±ëŠ¥ì„ í‰ê°€í•œë‹¤.
    * ì›ì‹œ ë°ì´í„°ëŠ” ğ¾ğ‘– ê°’ì´ ì—†ëŠ” ë‹¨ë°±ì§ˆ-ë¦¬ê°„ë“œ ìŒ ì œê±°, ì „ì²´ ë°ì´í„° ì§‘í•©ì˜ 80%ë¥¼ ë¬´ì‘ìœ„ë¡œ ì„ íƒí•˜ì—¬ í›ˆë ¨ ì§‘í•© êµ¬ì¶•, ë‚˜ë¨¸ì§€ 20%ëŠ” í‰ê°€ë¥¼ ìœ„í•´ ë…ë¦½ì ì¸ ì‹œí—˜ ì§‘í•©ìœ¼ë¡œ ì €ì¥ë˜ëŠ” ë“± ë¨¼ì € ì‚¬ì „ ì²˜ë¦¬ëœë‹¤.
* ë³¸ ë…¼ë¬¸ì—ì„œëŠ” BindingDB ë¥¼ ì‚¬ìš©í–ˆë‹¤
  * BindingDB: 2021ë…„ 7ì›” 29ì¼ê¹Œì§€ ì‹¤í—˜ì ìœ¼ë¡œ í™•ì¸ëœ 8,005ê°œì˜ í‘œì  ë‹¨ë°±ì§ˆê³¼ 986,143ê°œì˜ ì‘ì€ ì•½ë¬¼ ë¶„ì ì‚¬ì´ì˜ ê²°í•© ì¹œí™”ì„±ì„ í¬í•¨í•˜ëŠ” ì§€ì†ì ìœ¼ë¡œ ì—…ë°ì´íŠ¸ë˜ëŠ” ë°ì´í„° ì„¸íŠ¸
* ëª¨ë¸ ê°œë°œì„ ìœ„í•œ ë°ì´í„° ì„¸íŠ¸ êµ¬ì„±ì— ì ìš©í•œ ê¸°ì¤€
  * (1) excluding binding interactions with multichain protein complexes because it is not capable of identifying which chain of the protein interacts with the molecular; 
  * (2) retaining binding interactions only represented by ğ¾ğ‘– value and it means that other measurements in the form of ğ¼ğ¶50 or ğ¾ğ‘‘ values are removed; 
  * (3) keeping common drug molecules and target proteins occurring in at least three and six interactions in the entire dataset [11], respectively; 
  * (4) removing data with invalid ğ¾ğ‘– values. For example, we notice that some data used â€œ>â€ and â€œ<â€ in the ğ¾ğ‘– values to indicate ranges and we directly exclude them for the subsequent analysis. Additionally, there are some zeros in the ğ¾ğ‘– values which should not appear based on the definition of ğ¾ğ‘–. Thus, we treat them as invalid values and simply removed them;
  * (5) transferring ğ¾ğ‘– values into ğ‘ğ¾ğ‘– using the following equation (eq.1) due to the skewed distribution of original values of ğ¾ğ‘–: ğ‘ğ¾ğ‘– = âˆ’log<sub>10</sub> ğ¾ğ‘– --- (eq.1)
  * (6) removing data where the ğ‘ğ¾ğ‘– values are less than -8 because these extremely small ğ‘ğ¾ğ‘– values suggest that the corresponding drug molecules and protein targets held much weak or even no interactions. As a result, a total of 249,517 interactions with 54,135 drug molecules and 1,844 protein targets are finally used in developing our model.

![](2022-05-18-13-55-41.png)

* Mol2Vecì€ ë¶„ìë¥¼ ìˆ«ì ë²¡í„° í‘œí˜„ìœ¼ë¡œ ë³€í™˜í•˜ê¸° ìœ„í•œ ë¹„ì§€ë„ ë”¥ ëŸ¬ë‹ ê¸°ë°˜ ì ‘ê·¼ ë°©ì‹ì´ë‹¤. ìì—°ì–´ ì²˜ë¦¬(NLP) ê¸°ë²•ì— ì˜ê°ì„ ë°›ì•„, Mol2Vecì€ ëª¨ê±´ ì‹ë³„ì [23]ì— ì˜í•´ ì–»ì–´ì§„ ë¶„ì í•˜ìœ„ êµ¬ì¡°ë¥¼ "ë‹¨ì–´"ë¡œ ê°„ì£¼í•˜ê³ , í™”í•©ë¬¼ì„ "ë¬¸ì¥"ìœ¼ë¡œ ê°„ì£¼í•˜ê³ , ì´ë¥¼ ì— ê¸°ë°˜í•œ ì¡°ë°€í•œ ë²¡í„° í‘œí˜„ìœ¼ë¡œ ì¸ì½”ë”©í•œë‹¤.
ì´ë¥¸ë°” í™”í•©ë¬¼ ë§ë­‰ì¹˜ì…ë‹ˆë‹¤.

* ProSEëŠ” ë‹¨ë°±ì§ˆ êµ¬ì¡° ì •ë³´ë¥¼ ì¸ì½”ë”©í•˜ëŠ” ìˆ«ì ë²¡í„°ë¡œ ë‹¨ë°±ì§ˆ ì„œì—´ì„ í‘œí˜„í•˜ê¸° ìœ„í•´ ê°œë°œëœ ë”¥ ëŸ¬ë‹ ê¸°ë°˜ ë°©ë²•ì´ë‹¤. ë¨¼ì € ë‹¨ë°±ì§ˆ ì„œì—´ì„ ìœ ì‚¬í•œ ì•„ë¯¸ë…¸ì‚°(ë‹¨ì–´)ì„ ê°€ê¹Œìš´ ìˆ«ìë¡œ ë§¤í•‘í•˜ëŠ” íŠ¹ì • ì•ŒíŒŒë²³ ëª©ë¡("ë¬¸ì¥")ìœ¼ë¡œ ë²ˆì—­í•œë‹¤. ê·¸ëŸ° ë‹¤ìŒ, ProSE ëª¨ë¸ì€ ë‹¨ì–´ë¥¼ ìˆ«ì ë²¡í„°ë¡œ ì¸ì½”ë”©í•œë‹¤. 
* Mol2Vec(ë‹¤ìš´ë¡œë“œ ë§í¬: https://github.com/samoturk/mol2vec)
* ProSE(ë‹¤ìš´ë¡œë“œ ë§í¬, https://github.com/tbepler/prose)
  * í™œìš©í•˜ì—¬ ê°ê° ì•½ë¬¼ ë¶„ì í™”í•©ë¬¼ê³¼ ë‹¨ë°±ì§ˆì— ëŒ€í•´ ê³ ì •ëœ ê¸¸ì´ì˜ ë²¡í„° í‘œí˜„ì„ ì–»ëŠ”ë‹¤.
* PROSE ì‚¬ìš©ë„ ì‚¬ìš©ë²•ì— ê´€í•œ ì˜ˆì‹œê°€ ì—†ì–´ì„œ ì‚¬ìš©í•˜ê¸°ê°€ ì–´ë ¤ì›€.


## 6. Protein identificationì„ ìœ„í•œ bio.tools ì‚¬ì´íŠ¸
MSAC |   
The Molecular Spectrometry Adduct Calculator (MSAC) calculates the m/z of potential adducts from given compound m/z.

RCSB Protein Data Bank |   
Improved Annotation, Search, and Visualization of Membrane Protein Structures Archived in the PDB.

MS2Rescore |   
Sensitive PSM rescoring with predicted MSÂ² peak intensities using MSÂ²PIP, DeepLC, and Percolator.

PeptideShaker Online |   
A User-Friendly Web-Based Framework for the Identification of Mass Spectrometry-Based Proteomics Data.

PROMISed |   
A novel web-based tool to facilitate analysis and visualization of the molecular interaction networks from co-fractionation mass spectrometry (CF-MS) experiments.

PROSE |   
Single-sample proteome enrichment enables missing protein recovery and phenotype association.

RHybridFinder |   
An R package to process immunopeptidomic data for putative hybrid peptide discovery.

VIQoR |   
VIQoR, is a user-friendly web service for Visually supervised protein Inference and protein Quantification implemented in R. The Shiny web interface integrates the post-identification processes involved in protein inference and relative protein abundance summarization, along with smart and novel interactive visualization modules to support the common researchers with a straight-forward tool for protein quantification, data browsing and data inspection.

ProteInfer |   
ProteInfer is an approach for predicting the functional properties of protein sequences using deep neural networks.

prPred-DRLF |   
prPred-DRLF is a tool to identify the plant resistance proteins (R proteins) based on deep representation learning features

## 7. Reference: Learning the protein language: Evolution, structure, and function (Cell Systems 12, 654â€“669)

* ì–¸ì–´ ëª¨ë¸ì€ ìµœê·¼ ëŒ€ê·œëª¨ ë‹¨ë°±ì§ˆ ì„œì—´ ë°ì´í„°ë² ì´ìŠ¤ì—ì„œ ì •ë³´ë¥¼ ì¦ë¥˜í•˜ê¸° ìœ„í•œ ê°•ë ¥í•œ ê¸°ê³„ í•™ìŠµ ì ‘ê·¼ ë°©ì‹ìœ¼ë¡œ ë¶€ìƒí•˜ê³  ìˆë‹¤.
  * ì‰½ê²Œ êµ¬í•  ìˆ˜ ìˆëŠ” ì‹œí€€ìŠ¤ ë°ì´í„°ë§Œìœ¼ë¡œ, ì´ëŸ¬í•œ ëª¨ë¸ì€ ë‹¨ë°±ì§ˆ ê³µê°„ ì „ë°˜ì— ê±¸ì¹œ ì§„í™”ì , êµ¬ì¡°ì , ê¸°ëŠ¥ì  ì¡°ì§ì„ ë°œê²¬í•œë‹¤. 
  * ì–¸ì–´ ëª¨ë¸ì„ ì‚¬ìš©í•˜ì—¬ ì•„ë¯¸ë…¸ì‚° ì‹œí€€ìŠ¤ë¥¼ êµ¬ì¡°ì  ë° ê¸°ëŠ¥ì  íŠ¹ì„±ì„ í¬ì°©í•˜ëŠ” ë¶„ì‚° ë²¡í„° í‘œí˜„ìœ¼ë¡œ ì¸ì½”ë”©í•  ìˆ˜ ìˆì„ ë¿ë§Œ ì•„ë‹ˆë¼ ì‹œí€€ìŠ¤ ë³€í˜•ë“¤ì˜ ì§„í™”ì  ì í•©ì„±ì„ í‰ê°€í•  ìˆ˜ ìˆë‹¤. 
  * ë³¸ ë…¼ë¬¸ì—ì„œëŠ” ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸ë§ì˜ ìµœê·¼ ë°œì „ê³¼ ë‹¤ìš´ìŠ¤íŠ¸ë¦¼ ë‹¨ë°±ì§ˆ íŠ¹ì„± ì˜ˆì¸¡ ë¬¸ì œì— ëŒ€í•œ ì‘ìš©ì— ëŒ€í•´ ë…¼ì˜í•œë‹¤. ê·¸ëŸ° ë‹¤ìŒ ì´ëŸ¬í•œ ëª¨ë¸ì´ ì‚¬ì „ ìƒë¬¼í•™ì  ì§€ì‹ìœ¼ë¡œ í’ë¶€í•´ì§ˆ ìˆ˜ ìˆëŠ” ë°©ë²•ì„ ê³ ë ¤í•˜ê³  ë‹¨ë°±ì§ˆ êµ¬ì¡° ì§€ì‹ì„ í•™ìŠµëœ í‘œí˜„ìœ¼ë¡œ ì¸ì½”ë”©í•˜ê¸° ìœ„í•œ ì ‘ê·¼ ë°©ì‹ì„ ì†Œê°œí•œë‹¤. 
  * ì´ëŸ¬í•œ ëª¨ë¸ì— ì˜í•´ ì¦ë¥˜ëœ ì§€ì‹ì€ ì „ì´ í•™ìŠµì„ í†µí•´ ë‹¤ìš´ìŠ¤íŠ¸ë¦¼ í•¨ìˆ˜ ì˜ˆì¸¡ì„ ê°œì„ í•  ìˆ˜ ìˆê²Œ í•œë‹¤. 
  * ì‹¬ì¸µ ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸ì€ ë‹¨ë°±ì§ˆ ìƒë¬¼í•™ì— í˜ëª…ì„ ì¼ìœ¼í‚¤ê³  ìˆë‹¤. ê·¸ë“¤ì€ ë‹¨ë°±ì§ˆê³¼ ì¹˜ë£Œ ì„¤ê³„ì— ì ‘ê·¼í•˜ëŠ” ìƒˆë¡œìš´ ë°©ë²•ì„ ì œì•ˆí•œë‹¤. 
    * ê·¸ëŸ¬ë‚˜ ê°•ë ¥í•œ ìƒë¬¼í•™ì  ì‚¬ì „ë“¤ì„ ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸ë¡œ ì¸ì½”ë”©í•˜ê³  ë” ë„“ì€ ì»¤ë®¤ë‹ˆí‹°ì— ëŒ€í•œ ì ‘ê·¼ì„±ì„ ë†’ì´ê¸° ìœ„í•´ì„œëŠ” ì¶”ê°€ì ì¸ ê°œë°œì´ í•„ìš”í•˜ë‹¤.


  * ë‹¨ë°±ì§ˆì€ ì„¸í¬ì˜ ë¶„ì ê¸°ëŠ¥ì˜ ëŒ€ë¶€ë¶„ì„ ìˆ˜í–‰í•˜ëŠ” ë¶„ì ê¸°ê³„ì´ë‹¤.
    * ì´ë“¤ì€ 3ì°¨ì› êµ¬ì¡°ì˜ ë³µì¡í•œ ì•™ìƒë¸”ì— ì ‘íˆëŠ” ì•„ë¯¸ë…¸ì‚°ì˜ ì„ í˜• ë°°ì—´ë¡œ êµ¬ì„±ë˜ë©°, ìˆœì„œì—ì„œ ë¬´ì§ˆì„œí•œ ìƒíƒœê¹Œì§€ ë‹¤ì–‘í•˜ë©° ì…ì²´êµ¬ì¡° ë³€í™”ë¥¼ ê²ªì„ ìˆ˜ ìˆë‹¤.
    * ìƒí™”í•™ì , ì„¸í¬ì  ê¸°ëŠ¥ì€ ë‹¨ë°±ì§ˆ ì„œì—´ê³¼ êµ¬ì¡°ì—ì„œ ë‚˜íƒ€ë‚œë‹¤. ì„œì—´-êµ¬ì¡°-ê¸°ëŠ¥ ê´€ê³„ë¥¼ ì´í•´í•˜ëŠ” ê²ƒì€ ë‹¨ë°±ì§ˆ ìƒë¬¼í•™ì˜ í•µì‹¬ ë¬¸ì œì´ë©° ì§ˆë³‘ ë©”ì»¤ë‹ˆì¦˜ì„ ì´í•´í•˜ê³  ì¹˜ë£Œ ë° ìƒëª…ê³µí•™ ì‘ìš©ì„ ìœ„í•œ ë‹¨ë°±ì§ˆê³¼ ì•½ë¬¼ì„ ì„¤ê³„í•˜ëŠ” ë° ì¤‘ì¶”ì ì´ë‹¤. 
    * ì‹œí€€ìŠ¤-êµ¬ì¡°-í•¨ìˆ˜ ê´€ê³„ì˜ ë³µì¡ì„±ì€ ë¶€ë¶„ì ìœ¼ë¡œ ê¸°ì¡´ ë„êµ¬ê°€ ëŒ€ê·œëª¨ ë°ì´í„°ë² ì´ìŠ¤ì— ì €ì¥ëœ ì‹œí€€ìŠ¤, êµ¬ì¡° ë° ê¸°ëŠ¥ ì •ë³´ì˜ ì¦ê°€í•˜ëŠ” ì–‘ì˜ ì ì¬ë ¥ì„ ì™„ì „íˆ ì‹¤í˜„í•˜ì§€ ëª»í•˜ê¸° ë•Œë¬¸ì— ìš°ë¦¬ì˜ ê³„ì‚° ëª¨ë¸ë§ ëŠ¥ë ¥ì— ê³„ì† ë„ì „í•œë‹¤.
    * ìµœê·¼ê¹Œì§€ ë‹¨ë°±ì§ˆ ë¶„ì„ì„ ìœ„í•œ ê³„ì‚° ë°©ë²•ì€ ì§„í™”ì , ë”°ë¼ì„œ ê¸°ëŠ¥ì  ì••ë ¥ì„ ë°˜ì˜í•˜ëŠ” ì‹œí€€ìŠ¤ íŒ¨í„´ì„ ì‹ë³„í•˜ë ¤ëŠ” ìµœì´ˆì˜ ì›ë¦¬ ê¸°ë°˜ êµ¬ì¡° ì‹œë®¬ë ˆì´ì…˜ ë˜ëŠ” í†µê³„ì  ì‹œí€€ìŠ¤ ëª¨ë¸ë§ ì ‘ê·¼ë²•ì„ ì‚¬ìš©í•´ ì™”ë‹¤(Marks, Hopf and Sander, 2012; Ekeberg et al., 2013; Wang et al., 2017; Liu et). al. , 2018; Yang et al., 2020) (ê·¸ë¦¼ 1) ì´ëŸ¬í•œ ë°©ë²•ë“¤ ë‚´ì—ì„œ êµ¬ì¡° ë¶„ì„ì€ ì£¼ë¡œ ì²« ë²ˆì§¸ ì›ì¹™ì¸ ë°˜ë©´, ì‹œí€€ìŠ¤ ë¶„ì„ ë°©ë²•ì€ ì£¼ë¡œ ì§„í™” ê³¼ì •ì— ëŒ€í•œ ê°•ë ¥í•œ ê°€ì •ì„ í•˜ëŠ” í†µê³„ ì‹œí€€ìŠ¤ ëª¨ë¸ì„ ê¸°ë°˜ìœ¼ë¡œ í•˜ì§€ë§Œ ì´ìš© ê°€ëŠ¥í•œ ìì—° ì‹œí€€ìŠ¤ ì •ë³´ì˜ ì–‘ì´ ì¦ê°€í•¨ì— ë”°ë¼ ì ì  ë” ë°ì´í„° ì¤‘ì‹¬í™”ë˜ì—ˆë‹¤.
    * ë¬¼ë¦¬ ê¸°ë°˜ ì ‘ê·¼ë²•ì€ ëª¨ë“  ì›ì ì—ë„ˆì§€ í•¨ìˆ˜(Hornak et al., 2006; Hess et al., 2008; Alford et al., 2017) ë˜ëŠ” íœ´ë¦¬ìŠ¤í‹±ì„ ì‚¬ìš©í•œë‹¤.
      * ì£¼ì–´ì§„ êµ¬ì„±ì˜ ì—ë„ˆì§€ë¥¼ ì¶”ì •í•˜ê³  ìì—°ì ì¸ ì›€ì§ì„ì„ ì‹œë®¬ë ˆì´ì…˜í•˜ê¸° ìœ„í•´ ë‹¨ë°±ì§ˆì„ ìœ„í•´ ì„¤ê³„ë˜ì—ˆë‹¤(Rohlet et al., 2004). ì´ ë°©ë²•ë“¤ì€ ë§¤ë ¥ì ì¸ë°, ì™œëƒí•˜ë©´ ì´ ë°©ë²•ë“¤ì€ ì´ ì‹œìŠ¤í…œë“¤ì˜ ë¬¼ë¦¬í•™ì— ëŒ€í•œ ìš°ë¦¬ì˜ ê·¼ë³¸ì ì¸ ì´í•´ë¥¼ ëŒì–´ë‚´ê³  í•´ì„ ê°€ëŠ¥í•œ ê°€ì„¤ì„ ìƒì„±í•˜ê¸° ë•Œë¬¸ì´ë‹¤. 
        * ì‘ì€ ì¼ì •í•œ í¬ê¸°ì˜ ì—°ì† ì‹œí€€ìŠ¤ì™€ ê´€ë ¨ëœ ì ‘íŒ íŒŒí¸ì„ ê¿°ë§¨ Rosetta ë„êµ¬ëŠ” ë‹¨ë°±ì§ˆ ì ‘í˜ ë° ì„¤ê³„ë¥¼ ìœ„í•œ ììœ  ì—ë„ˆì§€ ì¶”ì •(Leaver-Fay et al., 2011)ì„ ì‚¬ìš©í•˜ëŠ” ë° í˜„ì €í•˜ê²Œ ì„±ê³µí–ˆìœ¼ë©°, GROMACSì™€ ê°™ì€ ë¶„ì ì—­í•™ ì†Œí”„íŠ¸ì›¨ì–´ëŠ” ì—­í•™ ëª¨ë¸ë§ê³¼ êµ¬ì¡° ì˜ˆì¸¡ ì„¸ë¶„í™”(Hess et al., 2008)ì— ë„ë¦¬ ì‚¬ìš©ëœë‹¤. 
        * ê±°ì¹œ ì…ì ì—ë„ˆì§€ í•¨ìˆ˜ë¥¼ ê¸°ë°˜ìœ¼ë¡œ ì ‘ê·¼ ê°€ëŠ¥í•œ êµ¬ì„±ì—ì„œ í‘œë³¸ì„ ì¶”ì¶œí•˜ëŠ” í†µê³„ ìƒ˜í”Œë§ ì ‘ê·¼ ë°©ì‹ë„ ê°œë°œë˜ì—ˆë‹¤(Godzik, Kolinski ë° Scolnick, 1993; Srinivasan and Rose, 1995; Choi and Pappu, 2019). RosettaëŠ” íŠ¹íˆ êµ¬ì¡° í…œí”Œë¦¿ê³¼ ììœ  ì—ë„ˆì§€ ìµœì†Œí™”ë¥¼ í˜¼í•©í•˜ì—¬ ëŒ€ìƒ êµ¬ì¡°ì™€ ì¼ì¹˜í•˜ëŠ” ì‹œí€€ìŠ¤ë¥¼ ì°¾ì•„ ì„¤ê³„ ë¬¸ì œë¥¼ í•´ê²°í•˜ëŠ” ë° ì„±ê³µí–ˆë‹¤. ê·¸ëŸ¬ë‚˜ Rosettaì˜ ì„±ê³µì—ë„ ë¶ˆêµ¬í•˜ê³ , Rosettaì™€ ìœ ì‚¬í•œ ì ‘ê·¼ ë°©ì‹ì€ ë‹¨ìˆœí™”ëœ ì—ë„ˆì§€ ëª¨ë¸ì„ ê°€ì •í•˜ê³ , ê³„ì‚° ë¹„ìš©ì´ ë§¤ìš° ë¹„ì‹¸ë©°, ì˜¬ë°”ë¥´ê²Œ ì„¤ì •í•˜ê¸° ìœ„í•´ ì „ë¬¸ ì§€ì‹ì´ í•„ìš”í•˜ë©°, ì •í™•ë„ê°€ ì œí•œì ì´ë‹¤.

ì•ŒíŒŒí´ë“œ2 ê³µë¶€í•´ë³¼ ê²ƒ!
  * ë‹¨ë°±ì§ˆì˜ ë¶„ì‚° ë²¡í„° í‘œí˜„ì´ ë‹¨ë°±ì§ˆ ì‹œí€€ìŠ¤ì˜ ìƒì„± ëª¨ë¸ì—ì„œ ì¶”ì¶œë  ìˆ˜ ìˆê³ , ìì—° ë‹¨ë°±ì§ˆ ê³µê°„ì— ê±¸ì³ í¬ê³  ë‹¤ì–‘í•œ ì‹œí€€ìŠ¤ì˜ ë°ì´í„°ë² ì´ìŠ¤ë¡œë¶€í„° í•™ìŠµë  ìˆ˜ ìˆìœ¼ë©°, ë”°ë¼ì„œ ì£¼ì–´ì§„ ì‹œí€€ìŠ¤ì˜ ì˜ë¯¸ ë˜ëŠ” ê¸°ëŠ¥ì„ í¬ì°©í•  ìˆ˜ ìˆë‹¤ëŠ” ìƒê°ì— ê¸°ì´ˆ
    * ê¸°ëŠ¥ì€ ë‹¨ë°±ì§ˆì´ í•˜ëŠ” ê²ƒê³¼ ê´€ë ¨ëœ ëª¨ë“  ì„±ì§ˆ
  * ìê¸° íšŒê·€ ì–¸ì–´ ëª¨ë¸ì—ì„œ, ì‹œí€€ìŠ¤ì˜ í™•ë¥ ì€ ê° í† í°ì˜ í™•ë¥ ì´ ì´ì „ í† í°ì—ì„œë§Œ ì¡°ê±´í™”ë˜ë„ë¡ ì¸ìˆ˜ ë¶„í•´ëœë‹¤.
    * ì´ ì¸ìˆ˜ ë¶„í•´ëŠ” ì •í™•í•˜ë©° ë¶„í¬ì—ì„œ í‘œë³¸ì„ ì¶”ì¶œí•˜ê±°ë‚˜ í™•ë¥  ìì²´ë¥¼ í‰ê°€í•˜ëŠ” ë° ìœ ìš©
    * ê³µì‹ì˜ ë‹¨ì ì€ ê° ìœ„ì¹˜ì— ëŒ€í•´ í•™ìŠµëœ í‘œí˜„ì´ ì´ì „ ìœ„ì¹˜ì—ë§Œ ì˜ì¡´í•˜ì—¬ ì ì¬ì ìœ¼ë¡œ ìƒí™©ë³„ í‘œí˜„ìœ¼ë¡œ ëœ ìœ ìš©
  * êµ¬ê¸€ì˜ BERTì™€ ì˜¤í”ˆAIì˜ GTP-3ëŠ” ê±°ëŒ€í•œ ì˜¨ë¼ì¸ ë§ë­‰ì¹˜ì˜ ìˆ˜ì‹­ì–µ ê°œì˜ í…ìŠ¤íŠ¸ í•­ëª©ì—ì„œ í•™ìŠµí•˜ëŠ” ëŠ¥ë ¥ì— ì˜í•´ ì£¼ë¡œ êµ¬ë™
  * ìµœê·¼ NLPì˜ ë°œì „ ì´ìœ 
    * neural network architectures, 
    * new training approaches, 
    * increasing compute power, and 
    * increasing accessibility of huge text corpuses. 
  * LSTMs are recurrent neural networks
    * í•œ ë²ˆì— í•˜ë‚˜ì˜ í† í° ì‹œí€€ìŠ¤ë¥¼ ìˆœì„œëŒ€ë¡œ ì²˜ë¦¬í•˜ë¯€ë¡œ ìœ„ì¹˜ì™€ ëª¨ë“  ì´ì „ ìœ„ì¹˜ì—ì„œ ì •ë³´ë¥¼ ìº¡ì²˜í•˜ëŠ” í‘œí˜„ì„ í•™ìŠµ
    * ì£¼ì–´ì§„ ìœ„ì¹˜ ì „í›„ì— í† í°ì˜ ì •ë³´ë¥¼ í¬í•¨í•˜ê¸° ìœ„í•´, ì–‘ë°©í–¥ LSTMì€ ê° ë ˆì´ì–´ì˜ ìˆœë°©í–¥ê³¼ ì—­ë°©í–¥ìœ¼ë¡œ ì‘ë™í•˜ëŠ” ë‘ ê°œì˜ ê°œë³„ LSTMì„ ê²°í•©
    * ì „ì²´ ì‹œí€€ìŠ¤ ì»¨í…ìŠ¤íŠ¸ë¥¼ í¬í•¨í•œ í‘œí˜„ì„ í•™ìŠµí•  ìˆ˜ ìˆì§€ë§Œ, ì‹¤ì œë¡œ ì›ê²© ì¢…ì†ì„±ì„ í•™ìŠµí•˜ëŠ” ëŠ¥ë ¥ì€ ì œí•œì 
    * í•œê³„ë¥¼ í•´ê²°í•˜ê¸° ìœ„í•´, ë³€ì••ê¸°ëŠ” ì‹œí€€ìŠ¤ì˜ ê° ìœ„ì¹˜ì— ëŒ€í•œ ì£¼ì˜ ë²¡í„°ë¥¼ ëª…ì‹œì ìœ¼ë¡œ ê³„ì‚°í•˜ì—¬ í‘œí˜„ì„ í•™ìŠµ
  * self-attention module 
    * ì‹œí€€ìŠ¤ì˜ ê° ìš”ì†Œì˜ ì¶œë ¥ í‘œí˜„ì€ ê°€ì¤‘ì¹˜ ìì²´ê°€ ì…ë ¥ì˜ í•™ìŠµëœ ë³€í™˜ì„ ê¸°ë°˜ìœ¼ë¡œ í•˜ëŠ” ê° ìœ„ì¹˜ì—ì„œ ì…ë ¥ í‘œí˜„ì˜ ë³€í™˜ì— ëŒ€í•œ ê°€ì¤‘ì¹˜ í•©ìœ¼ë¡œ ê³„ì‚°ë¨
  * attention mechanism 
    * ì¼ë°˜ì ìœ¼ë¡œ ë³€ì••ê¸°ê°€ ì„ í˜• ì‹œí€€ìŠ¤ì—ì„œ ë©€ë¦¬ ìˆëŠ” ìœ„ì¹˜ ê°„ì˜ ì¢…ì†ì„±ì„ ë” ì‰½ê²Œ í•™ìŠµí•  ìˆ˜ ìˆë„ë¡ í•¨
  * Transformers
    * autoregressive language modelsë¡œë„ ìœ ìš©
    * ìì—°ì–´ ì²˜ë¦¬ì—ì„œ í”¼í„°ìŠ¤ ë“±ì€ biLSTMì˜ ìˆ¨ê²¨ì§„ ê³„ì¸µ(ìŠ¤íƒ ì‹ ê²½ë§ì˜ ì¤‘ê°„ í‘œí˜„)ì´ ë¬¸ë§¥ì—ì„œ ë‹¨ì–´ì˜ ì˜ë¯¸ì  ì˜ë¯¸ë¥¼ ì¸ì½”ë”©í•œë‹¤ëŠ” ê²ƒì„ ì¸ì‹í–ˆë‹¤. ì´ ê´€ì°°ì€ ìƒë¬¼í•™ì  ì„œì—´ ë¶„ì„ì„ ìœ„í•´ ìƒˆë¡­ê²Œ í™œìš©
    * Although transformers are powerful models, they require enormous numbers of parameters and train more slowly than typical recurrent neural networks. 
  * ê²°ë¡  ë° ê´€ì 
  * ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸ì„ ê°œì„ í•˜ëŠ” ì—´ì‡  ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸ë§ ë° í‘œí˜„ í•™ìŠµì˜ í–¥í›„ ë°œì „ì€ ë‹¨ë°±ì§ˆì— ê³ ìœ í•œ íŠ¹ì„±ì„ ëª¨ë¸ë§í•´ì•¼ í•  ê²ƒ
  * ìƒë¬¼í•™ì  ì„œì—´ì€ ìì—° ì–¸ì–´ê°€ ì•„ë‹ˆë©°, ìš°ë¦¬ëŠ” ìƒë¬¼í•™ì  ì„œì—´ì˜ ê·¼ë³¸ì ì¸ ë³¸ì§ˆì„ í¬ì°©í•˜ëŠ” ìƒˆë¡œìš´ ì–¸ì–´ ëª¨ë¸ì„ ê°œë°œí•´ì•¼ í•¨

## 8. Reference: Protein embeddings and deep learning predict binding residues for various ligand classes (Scientifc Reports | (2021) 11:23916)

* ë‹¨ë°±ì§ˆ ê¸°ëŠ¥ì˜ í•œ ê°€ì§€ ì¤‘ìš”í•œ ì¸¡ë©´ì€ ì‘ì€ ë¶„ì, ê¸ˆì† ì´ì˜¨, DNA ë˜ëŠ” RNAì™€ ê°™ì€ ê³ ë¶„ìë¥¼ í¬í•¨í•˜ëŠ” ë¦¬ê°„ë“œì— ë‹¨ë°±ì§ˆì˜ ê²°í•©ì´ë‹¤. 
* ë³¸ ë…¼ë¬¸ì€ ë‹¨ë°±ì§ˆ ì”ê¸°ê°€ ê¸ˆì† ì´ì˜¨, í•µì‚° ë˜ëŠ” ì‘ì€ ë¶„ìì— ê²°í•©í•˜ëŠ”ì§€ ì˜ˆì¸¡í•˜ëŠ” ë°©ë²•ì¸ bindEmbed21ì„ ì œì•ˆ
* ì¸ê³µì§€ëŠ¥(AI) ê¸°ë°˜ ë°©ë²•ì€ íŠ¸ëœìŠ¤í¬ë¨¸ ê¸°ë°˜ ë‹¨ë°±ì§ˆ ì–¸ì–´ ëª¨ë¸(pLM) ProtT5ì˜ ì„ë² ë”©ì„ ì…ë ¥ìœ¼ë¡œ ë…ì ì ìœ¼ë¡œ ì‚¬ìš©
* ë‹¤ì¤‘ ì‹œí€€ìŠ¤ ì •ë ¬(MSA)ì„ ìƒì„±í•˜ì§€ ì•Šê³  ë‹¨ì¼ ì‹œí€€ìŠ¤ë§Œ ì‚¬ìš©í•˜ì—¬ bindEmbed21DLì€ MSA ê¸°ë°˜ ì˜ˆì¸¡ì„ ëŠ¥ê°€í•¨
* ê°€ì¥ ê°•í•˜ê²Œ ì˜ˆì¸¡ëœ 25% ê²°í•© ì”ê¸°ì˜ ê²½ìš°, ì‹¤í—˜ ì£¼ì„ ëˆ„ë½ ë¬¸ì œë¥¼ ë¬´ì‹œí•˜ë”ë¼ë„ ì ì–´ë„ 73%ê°€ ì •í™•í•˜ê²Œ ì˜ˆì¸¡ë¨
* êµ¬ì¡°ë‚˜ MSAë¥¼ ì‚¬ìš©í•˜ì§€ ì•Šê³  ë¹ ë¥´ê³  ë‹¨ìˆœí•˜ë©° ê´‘ë²”ìœ„í•˜ê²Œ ì ìš©ê°€ëŠ¥

* ì¶•ì•½ì–´
  * AI Artifcial intelligence (expanding ML through deep learning, i.e., using more free parameters)
  * BFD Big Fantastic Database (large database of protein sequences)
  * CI Confdence interval
  * CNN Convolutional Neural Network
  * HBI Homology-based inference
  * (p)LM (Protein) language model
  * MCC Matthews Correlation Coefcient
  * ML Machine learning
  * MSA Multiple sequence alignment
  * PDB Protein Data Bank
  * PIDE Pairwise sequence identity
  * SOTA State-of-the-art
  * SVM Support vector machine

* AlphaFold 2ëŠ” ë‹¤ì¤‘ ì‹œí€€ìŠ¤ ì •ë ¬(MSA)ì˜ ì •ë³´ì— í¬ê²Œ ì˜ì¡´í•œë‹¤. MSAê°€ ì—†ëŠ” ìµœê·¼ì˜ êµ¬ì¡° ì˜ˆì¸¡ì€ ëœ ì •í™•í•˜ë‹¤. êµ¬ì¡° ì˜ˆì¸¡ì´ ê²°í•© ì”ë¥˜ë¬¼ì—ì„œ ê²°í•© ë¶€ìœ„ë¡œì˜ ë‹¨ê³„ë¥¼ ë„˜ì–´ ê²°í•© ì˜ˆì¸¡ì„ ì–´ëŠ ì •ë„ê¹Œì§€ ê°œì„ í•  ìˆ˜ ìˆì„ì§€ëŠ” ì—¬ì „íˆ ë¶ˆë¶„ëª…í•˜ë‹¤.
* bindEmbed21: 

## 9. Reference: Biological structure and function emerge from scalingunsupervised learning to 250 million protein sequences(PNAS2021 Vol. 118 No. 15 e2016239118)

* ë‹¨ë°±ì§ˆ ì„œì—´ì€ ìì—°ì–´ì™€ ë§¤ìš° ë‹¤ë¥¸ ê³¼ì •ìœ¼ë¡œë¶€í„° ë°œìƒí•œë‹¤. ëª¨ë¸ê³¼ ëª©ì  í•¨ìˆ˜ê°€ ë„ë©”ì¸ ê°„ì˜ ì°¨ì´ì— ê±¸ì³ ìì—°ì–´ ì „ë‹¬ì— íš¨ê³¼ì ì¸ì§€ëŠ” ë¶ˆí™•ì‹¤í•˜ë‹¤. ìš°ë¦¬ëŠ” ì§„í™” ë°ì´í„°ì— ëŒ€í•œ ê³ ìš©ëŸ‰ íŠ¸ëœìŠ¤í¬ë¨¸ ì–¸ì–´ ëª¨ë¸ì„ í›ˆë ¨í•¨ìœ¼ë¡œì¨ ì´ ì§ˆë¬¸ì„ íƒêµ¬í•œë‹¤. ìš°ë¦¬ëŠ” ìƒë¬¼í•™ì  ì¡°ì§ ì›ë¦¬ì™€ ë³¸ì§ˆì ì¸ ìƒë¬¼í•™ì  íŠ¹ì„±ì— ëŒ€í•œ ì •ë³´ì˜ ì¡´ì¬ì— ëŒ€í•œ ê²°ê³¼ì ì¸ ë¹„ì§€ë„ í‘œí˜„ì„ ì¡°ì‚¬í•œë‹¤. ìš°ë¦¬ëŠ” ë¬¼ë¦¬í™”í•™ì  ê·œëª¨ì—ì„œ ì›ê²© ìƒë™í•™ì— ì´ë¥´ê¸°ê¹Œì§€ ì¡°ì§ ì›ë¦¬ì— ë¶€í•©í•˜ëŠ” ë©”íŠ¸ë¦­ êµ¬ì¡°ë¥¼ í‘œí˜„ ê³µê°„ì—ì„œ ë°œê²¬í•œë‹¤. ìš°ë¦¬ëŠ” ë˜í•œ í‘œí˜„ì—ì„œ 2ì°¨ ë° 3ì°¨ ë‹¨ë°±ì§ˆ êµ¬ì¡°ê°€ ì‹ë³„ë  ìˆ˜ ìˆë‹¤ëŠ” ê²ƒì„ ë°œê²¬í–ˆë‹¤. í‘œí˜„ì— ì˜í•´ ìº¡ì²˜ëœ êµ¬ì¡°ì  íŠ¹ì„±ì€ ì—¬ëŸ¬ í´ë”ì— ê±¸ì³ ì¼ë°˜í™”ëœë‹¤. ìš°ë¦¬ëŠ” í‘œí˜„ì„ ë‹¤ì–‘í•œ ì˜ˆì¸¡ ì‘ì—…ì— ì ìš©í•˜ê³  ì‘ìš© í”„ë¡œê·¸ë¨ ì „ì²´ì—ì„œ ìµœì²¨ë‹¨ ê¸°ëŠ¥ì„ í–¥ìƒì‹œí‚¨ë‹¤ëŠ” ê²ƒì„ ë°œê²¬í–ˆë‹¤.

* ì‹œí€€ìŠ¤ ë°ì´í„°ì—ì„œ ìƒë¬¼í•™ì  íŠ¹ì„±ì„ ë°°ìš°ëŠ” ê²ƒì€ ìƒë¬¼í•™ì„ ìœ„í•œ ìƒì„± ë° ì˜ˆì¸¡ ì¸ê³µì§€ëŠ¥ì„ í–¥í•œ ë…¼ë¦¬ì  ë‹¨ê³„ì´ë‹¤. ì—¬ê¸°ì„œëŠ” ë¹„ì§€ë„ í•™ìŠµì„ í†µí•´ ì‹¬ì¸µ ë¬¸ë§¥ ì–¸ì–´ ëª¨ë¸ì„ ì§„í™”ì  ë‹¤ì–‘ì„±ì— ê±¸ì¹œ ì‹œí€€ìŠ¤ë¡œ í™•ì¥í•  ê²ƒì„ ì œì•ˆí•œë‹¤. ìš°ë¦¬ëŠ” ì‚¬ì „ ì§€ì‹ ì—†ì´ 2ì°¨ êµ¬ì¡°, ì ‘ì´‰ ë° ìƒë¬¼í•™ì  í™œë™ê³¼ ê°™ì€ ë‹¨ë°±ì§ˆì˜ ê¸°ë³¸ íŠ¹ì„±ì— ëŒ€í•œ í•™ìŠµëœ í‘œí˜„ì—ì„œ ì •ë³´ê°€ ë‚˜íƒ€ë‚œë‹¤ëŠ” ê²ƒì„ ë°œê²¬í–ˆë‹¤. ìš°ë¦¬ëŠ” í•™ìŠµëœ í‘œí˜„ì´ ì›ê²© í˜¸ëª°ë¡œì§€ ê°ì§€, 2ì°¨ êµ¬ì¡° ì˜ˆì¸¡, ì¥ê±°ë¦¬ ì”ë¥˜-ì”ë¥˜ ì ‘ì´‰ ë° ëŒì—°ë³€ì´ íš¨ê³¼ë¥¼ ìœ„í•œ ë²¤ì¹˜ë§ˆí¬ì—ì„œ ìœ ìš©í•˜ë‹¤ëŠ” ê²ƒì„ ë³´ì—¬ì¤€ë‹¤. ë¹„ì§€ë„ í‘œí˜„ í•™ìŠµì€ ëŒì—°ë³€ì´ íš¨ê³¼ì™€ 2ì°¨ êµ¬ì¡°ì— ëŒ€í•œ ìµœì²¨ë‹¨ ì§€ë„ ì˜ˆì¸¡ì„ ê°€ëŠ¥í•˜ê²Œ í•˜ê³  ì¥ê±°ë¦¬ ì ‘ì´‰ ì˜ˆì¸¡ì„ ìœ„í•œ ìµœì²¨ë‹¨ ê¸°ëŠ¥ì„ í–¥ìƒì‹œí‚¨ë‹¤.
* Recently, self-supervision(ìê¸°ì§€ë„í•™ìŠµ) has emerged as a core direction inartificial intelligence research.
  * ë³¸ ë…¼ë¬¸ì€ ë‹¤ì–‘í•œ ìì—°ì–´ ì²˜ë¦¬ ì‘ì—…ì—ì„œ ìµœì²¨ë‹¨ ì„±ëŠ¥ì„ ì…ì¦í•œ ìì²´ ì§€ë„ ì–¸ì–´ ëª¨ë¸ë§ ì ‘ê·¼ ë°©ì‹ì„ íƒìƒ‰í•˜ì—¬ ë ˆì´ë¸”ì´ ì§€ì •ë˜ì§€ ì•Šì€ ì•„ë¯¸ë…¸ì‚° ì‹œí€€ìŠ¤ì˜ í˜•íƒœë¡œ ë‹¨ë°±ì§ˆ ë°ì´í„°ì— ì ìš©
  * ë‹¨ë°±ì§ˆ ì‹œí€€ìŠ¤ëŠ” 20ê°œì˜ í‘œì¤€ ìš”ì†Œì˜ ì‘ì€ ì–´íœ˜ë¥¼ ì‚¬ìš©í•˜ê¸° ë•Œë¬¸ì—, ëª¨ë¸ë§ ë¬¸ì œëŠ” ë‹¨ì–´ ìˆ˜ì¤€ ëª¨ë¸ë³´ë‹¤ ë¬¸ì ìˆ˜ì¤€ ì–¸ì–´ ëª¨ë¸(30, 31)ê³¼ ë” ìœ ì‚¬
  * ìì—°ì–´ì²˜ëŸ¼ ë‹¨ë°±ì§ˆ ì„œì—´ë„ ì¥ê±°ë¦¬ ì˜ì¡´ì„±ì„ í¬í•¨í•˜ë©°, ì›ê±°ë¦¬ ì»¨í…ìŠ¤íŠ¸ë¥¼ ê°ì§€í•˜ê³  ëª¨ë¸ë§í•˜ëŠ” ì•„í‚¤í…ì²˜ì˜ ì‚¬ìš©ì„ ë™ê¸° ë¶€ì—¬í•¨
* We investigate the Transformer (32), which has emerged as a powerful general purpose model architecture for representation learning and generative modeling, outperforming recurrent and convolutional architectures in natural language settings. 

[ESM](https://github.com/facebookresearch/esm)  
[ESM Embedding ì„¤ëª…](https://github.com/facebookresearch/esm/blob/main/examples/sup_variant_prediction.ipynb)

```
& python3 scripts/extract.py esm1b_t33_650M_UR50S examples/data/some_proteins.fasta examples/data/some_proteins_emb_esm1b/ \
    --repr_layers 0 32 33 --include mean per_tok
```
```python
scripts/extract.py
# extract.py íŒŒì¼ì˜ ì½”ë“œ ì‚¬ìš©ë¨
# scripts/extract.py has flags that determine what's included in the .pt file

esm1b_t33_650M_UR50S
# ì‚¬ì „ì— í•™ìŠµëœ ìë£Œ? ëª¨ë¸?

examples/data/some_proteins.fasta
# fasta file (sequence ì…ë ¥ëœ íŒŒì¼)

examples/data/some_proteins_emb_esm1b/ íŒŒì¼ìƒì„±
# ì½”ë“œê°€ ì‹¤í–‰ë˜ë©´ examples/data/some_proteins_emb_esm1b/ íŒŒì¼ì— ê²°ê³¼ íŒŒì¼ ìƒì„±
# ë””ë ‰í† ë¦¬ some_directory_filename_esm1b/ì— FASTA ì‹œí€€ìŠ¤ë‹¹ í•˜ë‚˜ì˜ .pt íŒŒì¼ì´ í¬í•¨ë˜ì–´ ìˆìŠµë‹ˆë‹¤. torch.load toë¥¼ ì‚¬ìš©í•˜ì—¬ íŒŒì¼ì„ ë¡œë“œí•˜ë©´ ë¨.

```

* FASTA formatì€ ìƒë¬¼ ì •ë³´í•™ ë° ìƒí™”í•™ì—ì„œ nucleotide sequence ë˜ëŠ” amino acid (protein) sequenceë¥¼ ë‚˜íƒ€ë‚´ëŠ” text-based formatìœ¼ë¡œ nucleotide ë˜ëŠ” amino acidsëŠ” ë‹¨ì¼ ë¬¸ì ì½”ë“œ (one-letter-code)ë¥¼ ì‚¬ìš©í•˜ì—¬ í‘œí˜„

* ì˜ˆì‹œ

```
>gi|5524211|gb|AAD44166.1| cytochrome b [Elephas maximus maximu]
LCLYTHIGRNIYYGSYLYSETWNTGIMLLLITMATAFMGYVLPWGQMSFWGATVITNLFSAIPYIGTNLV
EWIWGGFSVDKATLNRFFAFHFILPFTMVALAGVHLTFLHETGSNNPLGLTSDSDKIPFHPYYTIKDFLG
LLILILLLLLLALLSPDMLGDPDNHMPADPLNTPLHIKPEWYFLFAYAILRSVPNKLGGVLALFLSIVIL
GLMPFLHTSKHRSMMLRPLSQALFWTLTMDLLTLTWIGSQPVEYPYTIIGQMASILYFSIILAFLPIAGX
IENY
```
* NCBIëŠ” ìƒëª…ê³¼í•™ ë° ì˜í•™ ë…¼ë¬¸ ì¸ë±ìŠ¤ì˜ ë°ì´í„°ë² ì´ìŠ¤ì¸ íë©”ë“œ(PubMed), ìœ ì „ì²´ ì„œì—´ ë°ì´í„°ë² ì´ìŠ¤ì¸ ì§„ë±…í¬(GenBank)ë¥¼ ë¹„ë¡¯í•˜ì—¬ ê°ì¢… ìƒëª…ê³µí•™ ì •ë³´ë“¤ì„ ë‹´ê³  ìˆìœ¼ë©°, ì´ ëª¨ë“  ì •ë³´ë“¤ì€ Entrez ê²€ìƒ‰ì—”ì§„ì„ ì´ìš©í•˜ì—¬ ì˜¨ë¼ì¸ìœ¼ë¡œ ì—´ëŒí•  ìˆ˜ ìˆë‹¤.

* ~.pt ëŠ” pytorch ëª¨ë¸ íŒŒì¼ì˜ í™•ì¥ìì´ë‹¤.
* ESM ê´€ë ¨ ë…¼ë¬¸ì€ ë‚˜ì¤‘ì— ë‹¤ì‹œ êµ¬í˜„í•´ë³´ê¸°.

## 10. ê¹ƒí—ˆë¸Œë¡œ ë„ì»¤ ì´ë¯¸ì§€ ì˜¬ë¦¬ê¸°

Github.com Packages Container Registry

ê¹ƒí—ˆë¸Œ íŒ¨í‚¤ì§€ì—ì„œ ë„ì»¤ëŠ” ì•ˆì“°ì´ê³  ì»¨í…Œì´ë„ˆ ì‚¬ìš©í•˜ë©´ ëœë‹¤.
![](2022-05-30-09-52-57.png)

1. PAT(Personal access tokens) ë°œê¸‰ë°›ê¸°
   ëª…ë ¹ì–´ë¥¼ ì´ìš©í•´ì„œ íŒ¨í‚¤ì§€ë¥¼ ì œì–´í• ìˆ˜ ìˆë„ë¡  
   write-packages  
   delete-packages  
   ë‘ê°€ì§€ë¥¼ on ì‹œí‚¤ê³  ë°œê¸‰ë°›ì€ í† í°ì€ ì €ì¥í•´ì•¼í•¨.

   
2. ë„ì»¤ë¥¼ ê¹ƒí—ˆë¸Œë¡œ ë¡œê·¸ì¸
   docker login ghcr.io -u username -p ~~ 
   
   ghcr.io : ê¹ƒí—ˆë¸Œ ë ˆì§€ìŠ¤íŠ¸ë¦¬ ì£¼ì†Œë¥¼ ì…ë ¥  
   -u ë’¤ëŠ” ID(github username)ì´ë‹¤. (ID: egoing)
   -p ë’¤ : í† í° ì…ë ¥
   í† í°ê¹Œì§€ ì…ë ¥í•˜ë©´ íˆìŠ¤í† ë¦¬ì— ë‚¨ê¸°ë•Œë¬¸ì— ë‹¤ë¥¸ì‚¬ëŒë„ ë¡œê·¸ì¸ ê°€ëŠ¥ì„±ì´ ìˆê¸°ë•Œë¬¸

   ê·¸ë˜ì„œ export CR_PAT ì‚¬ìš© &rArr; í™˜ê²½ë³€ìˆ˜ì— í† í° ì €ì¥

```bach
export CR_PAT = ~~(PAT ì…ë ¥)
echo $CR_PAT
echo $CR_PAT | docker login ghcr.io -u username --password-stdin
```
--password-stdin ëŠ” CR_PAT ê°’ì´ë‹¤.  
ì‹¤í–‰ë§ˆë‹¤ í•´ì•¼í•˜ëŠ” ë¶ˆí¸í•¨ì„ ì—†ì• ë ¤ë©´  
shellì´ ì‹¤í–‰ë ë•Œë§ˆë‹¤ ì‚¬ìš©í•  ìˆ˜ ìˆë„ë¡ ì„¤ì •í•˜ë©´ëœë‹¤.

3. ë„ì»¤ ì»¨í…Œì´ë„ˆë¥¼ ì´ë¯¸ì§€ë¡œ ì €ì¥í•˜ëŠ” ë°©ë²•

ì»¨í…Œì´ë„ˆ ë³€ê²½ ì‚¬í•­ í™•ì¸

```
docker diff <CONTAINER_ID>
```
ì»¨í…Œì´ë„ˆ ì €ì¥

```
docker commit <CONTAINER_ID> <IMAGE_NAME>:<TAG>
 ```

Docker Hubì— ì´ë¯¸ì§€ Push

```
# ë ˆì§€ìŠ¤íŠ¸ë¦¬ì— ë¡œê·¸ì¸
docker login

# push
docker push <IMAGE_NAME>:<TAG>
```

githubì— ì´ë¯¸ì§€ Pushë°©ë²•

```
docker commit db_dpcr ghcr.io/choidb/cproject_db:1.0
docker push ghcr.io/choidb/cproject_db:1.0
```


